addons/l10n_vn_edi/models/sinvoice.py

```diff
@@ -0,0 +1,129 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+from collections import defaultdict
+
+from odoo import api, fields, models, _
+from odoo.exceptions import UserError
+
+
+# Invoice template that needs to be passed to Sinvoice and will determine the format of the resulting
+# invoice pdf on their system
+class SInvoiceTemplate(models.Model):
+    _name = 'sinvoice.template'
+    _description = 'SInvoice template'
+
+    name = fields.Char(
+        string='Template Code',
+        required=True,
+    )
+    template_invoice_type = fields.Selection(
+        selection=[
+            # Circular 32 being deprecated, only display types according to the Circular 78
+            ('1', '1 - Value-added invoice'),
+            ('2', '2 - Sales invoice'),
+            ('3', '3 - Public assets sales'),
+            ('4', '4 - National reserve sales'),
+            ('5', '5 - Invoice for national reserve sales'),
+            ('6', '6 - Warehouse release note'),
+        ],
+        required=True,
+    )
+    invoice_symbols_ids = fields.One2many(
+        comodel_name='sinvoice.symbol',
+        inverse_name='invoice_template_id',
+    )
+
+    _sql_constraints = [
+        ('name_uniq', 'unique (name)', 'The template code must be unique!')
+    ]
+
+    @api.constrains('name', 'template_invoice_type')
+    def _constrains_changes(self):
+        """
+        Multiple API endpoints will use these data, we should thus not allow changing them if they have been used
+        for any invoices sent to sinvoice.
+        """
+        # As we don't store the template on the invoice, we will first read group by symbol then group by template.
+        invoice_counts = self.env['account.move']._read_group(
+            domain=[
+                ('l10n_vn_edi_invoice_symbol.invoice_template_id', 'in', self.ids),
+                ('l10n_vn_edi_state', 'not in', ('ready_to_send', False)),  # Only matches sent invoices.
+            ],
+            groupby=['l10n_vn_edi_invoice_symbol'],
+            aggregates=['__count'],
+        )
+        invoices_per_template = defaultdict(int)
+        for symbol, count in invoice_counts:
+            invoices_per_template[symbol.invoice_template_id.id] += count
+
+        for record in self:
+            if invoices_per_template[record.id] > 0:
+                raise UserError(_('You cannot change the code or type of the template %s because it has '
+                                  'already been used to send invoices.', record.name))
+
+
+# Invoice symbol that needs to be passed to Sinvoice and will determine the prefix of the
+# invoice number on their system
+class SInvoiceSymbol(models.Model):
+    _name = 'sinvoice.symbol'
+    _description = 'SInvoice symbol'
+    """
+    The invoice symbols are made of multiple parts.
```

Comment:
is it worth it?
In any case, this data has to be set on the EDI platform too where the format gets validated already.
I believe it's less maintenance for us if we don't limit this since the format could change in the future (although it would render this comment out of date, it's mostly there to help devs if needed)